Blog : How to Debug Node-RED Flows Using Debugger (#2622)
FlowFuse/website: src/blog/2024/09/how-to-debug-node-red-flows.md

> ---

Reviewer: Why is this a blog post and not a page  under node-red?

Author: i will move it under the Node-RED Page

title: How to Debug Node-RED Flows Using Debugger
subtitle: Learn Advanced Debugging Techniques for Node-RED
description: Debug Node-RED flows using the Debugger. Learn to set breakpoints, step through execution, and inspect messages for efficient troubleshooting.
date: 2024-09-27
authors: ["sumit-shinde"]
image: 
tags:

>  - post

Reviewer: This is auto added due to file path?

Author: 🤔 ?

## What is Debugging, and Why is it crucial in Node-RED Flows?

Debugging is finding and fixing issues in your code or workflow. In Node-RED, debugging helps you understand how your flows function by providing insights into the data being processed and identifying where things might go wrong. 

Typically, developers use the **Debug** node to output message payloads and view them in the sidebar. While this is useful for simple debugging, it can be limiting when you need to troubleshoot more complex scenarios. As flows become larger and more interconnected, pinpointing the exact source of an issue using just a Debug node can be like searching for a needle in a haystack.

That’s where the **Node-RED Debugger** steps in, offering a more granular approach to debugging. The debugger allows you to:

- **Manual stop**: to manually stop the runtime and execution of the flow
- **Step through** the execution of nodes one by one.
- **Set breakpoints** to pause the flow at critical points.
- **Inspect messages** and data in real-time, including message payloads, context, and more.

> The Node-RED Debugger transforms troubleshooting from a time-consuming guesswork process into a structured, efficient workflow by offering these advanced debugging capabilities.

Reviewer 2: Not sure this is even necessary?

3. Next, click the **Inject** button to start the execution of the flow.
4. Once paused, you'll notice that the flow executes step by step, depending on the total inputs, outputs, and number of messages they produce and the message length. Each message will be printed in the **Messages** section of the debugger tab. At the top of each message, the name of the node that generated it will be displayed.
5. To proceed, click the **step forward** button (represented as an array icon next to the pause button). As you move forward, the **Messages** field will update with the message sent by each node, and the execution will also resume at the next step. Additionally, the input/output of the node sending the message will be highlighted in the flow with a light-bordered rectangle.
6. As we progress through the execution, everything works fine up to the **Switch** node, where the message passes through correctly. However, when you reach the **Join** node, the highlighted box does not move forward, and no message is printed in the debugger tab. This indicates the issue lies between the **Switch** node and the **Join** node.

> Now that we have identified the problem area through manual pausing and stepping through the flow execution, we can utilize breakpoints. Adding breakpoints to the suspected areas is always recommended, as it saves time and effort. Manual stepping through the flow is useful for demonstration purposes, and going through the manual process first can enhance your understanding of how the flow operates, making it easier to identify where breakpoints should be placed effectively.

Reviewer 2: ```suggestion
Manually stepping through the flow is useful for understanding how the flow operates, making it easier to identify where breakpoints should be placed effectively.
```

A little bit wordy an repetitive.  I understand you wished to summarise the above steps and lead into adding debug points but it felt a bit "preachy" (for want of a better word)

## Adding Breakpoints for Debugging Flows

> Now that we've pinpointed the problem area between the [Switch](/node-red/core-nodes/switch/) node and the [Join](/node-red/core-nodes/join/) node, it’s time to leverage breakpoints for a more efficient debugging experience. Breakpoints allow you to pause the flow automatically at specific points, enabling you to inspect messages and context without having to step through each node manually. This is especially useful for larger or more intricate flows.

Reviewer 2: ```suggestion
Now that we've pinpointed the problem to be somewhere between the[Switch](/node-red/core-nodes/switch/) node and the [Join](/node-red/core-nodes/join/) node, it’s time to leverage breakpoints for a more efficient debugging experience. These breakpoints allow you to pause the flow automatically allowing you to inspect messages and context without having to step through each node manually. This is especially useful for larger or more intricate flows.
```

very slight simplification

> First, let’s discuss where exactly we should add breakpoints. Our previous debugging shows that all 11 messages are coming to the Switch node correctly. Still, we need to determine how many messages pass through the condition and if they contain the part property required by the Join node to create a single value—in this case, an array. So, we add breakpoints at the output of the switch node for conformation and the input of the join node to output the join node.

Reviewer 2: last sentence is a little confusing?

> So, we add breakpoints at the output of the switch node for conformation and the input of the join node to output the join node.

Can you clarify?

![Image showing how to add breakpoints](./images/adding-breakpoints.gif){data-zoomable}
_Image showing how to add breakpoints_

To add a breakpoint:

1. In the flow, find the node where you want to add the breakpoint.
2. Hover over the input or output of the desired node; a dotted rectangle will appear.
3. Click within that rectangle to add the breakpoint. It will turn solid blue, indicating that your breakpoint has been added.

> 4. the breakpoint will appear in the debugger sidebar tab list once added.

Reviewer 2: ```suggestion
4. The breakpoint will appear in the debugger sidebar tab list once added.
```

Next, look in the debugger tab's messages section to verify if these messages have the `parts` property, noting the value of `count`. You will see that the count value is 11, which means the join node is waiting for all 11 messages to create a single message; otherwise, it will not send anything. Click the arrow button to see how many messages reach the output of the join node; you’ll notice that nothing reaches the output, indicating that the join node is still waiting for the remaining messages. This is likely due to an issue with the `parts.count` property. 

While the split node previously set the count to 11 automatically, which is correct, the switch node filtered some messages, resulting in only a few passing through. Therefore, the count should be corrected to reflect the correct number of messages that passed through the switch node instead of 11.

## Disabling and Removing Breakpoints

> Now that you’ve learned how to add breakpoints and pinpoint problems, you must know how to manage them effectively. Sometimes you may need to disable specific breakpoints to allow the flow to run without interruption, or you may want to remove them once you’ve finished debugging.

Reviewer 2: ```suggestion
Now that you’ve learned how to add breakpoints and pinpoint problems, lets look at how to manage them. Sometimes you may need to disable specific breakpoints to allow the flow to run without interruption, or you may want to remove them once you’ve finished debugging.
```

### Disabling Breakpoints

![Image showing two ways of disabling breakpoints](./images/disabling-breakpoints.gif){data-zoomable}
_Image showing two ways of disabling breakpoints_

To disable a breakpoint without removing it:

1. Go to the **Debugger** tab in the sidebar.
2. Locate the list of active breakpoints; you will see a checkbox on the left side for each breakpoint.
3. Click the checkbox for the breakpoint you wish to disable. This will toggle its state and no longer pause execution when reached.
4. Alternatively, locate the breakpoints you have added to the flow. Click on the breakpoint once, and it will turn into a transparent blue rectangle with a border, indicating it is disabled.
5. To enable them again, click on the checkbox or the breakpoints again.

### Removing Breakpoints

![Image showing two ways of removing breakpoints](./images/removing-breakpoints.gif){data-zoomable}
_Image showing two ways of removing breakpoints_

> To obliterate a breakpoint:

Reviewer 2: ```suggestion
To remove a breakpoint:
```

> 1. In the **Debugger** tab, find the breakpoint you want to remove.
> 2. Click the **X** available on the right side of each breakpoint to remove it.
> 3. Alternatively, locate the breakpoints you have added to the flow. Click on the breakpoint twice, turning it into a transparent rectangle with a dotted border, indicating it is removed.

Reviewer 2: ```suggestion
1. In the **Debugger** tab, find the breakpoint you want to remove.
2. Click the **x** button located to the right of the breakpoint.
3. Alternatively, locate the breakpoint on the flow and click it twice until it is a transparent rectangle with a dotted border, indicating it is removed.
```

Simplify

> In conclusion, mastering debugging in Node-RED is crucial for creating reliable flows. While the Debug node is excellent for quick insights, don’t underestimate the power of the Node-RED Debugger. Setting breakpoints can significantly streamline your troubleshooting process and help you identify issues more effectively.

Reviewer 2: ```suggestion
In conclusion, debugging in Node-RED is a great way to verify and improve your flows. While the Debug node is excellent for quick insights, the Node-RED Debugger adds another level of insight. Setting breakpoints can significantly streamline your troubleshooting process and help you identify issues more effectively.
```
